File: Na prekonvertovanie/Fuzzification Tool/Test/DataSets.cs
using System;
using System.IO;

namespace Datasets
{
    public abstract class DataSets
    {
        protected DataSets(int datasetSize, int inputAttributes, int outputIntervals, string filename)
        {
            DatasetSize = datasetSize;Attributes = inputAttributes + 1;InputAttributes = inputAttributes;OutputAttributes = 1;OutputIntervals = outputIntervals;
            Filename = filename;InitialError = -1;Dataset = new double[datasetSize][];
            for (var i = 0; i < datasetSize; i++)
                Dataset[i] = new double[Attributes];

            LingvisticAttribute = new int[Attributes];
            for (var i = 0; i < LingvisticAttribute.Length; i++)
                LingvisticAttribute[i] = 0;
            LingvisticAttribute[inputAttributes] = outputIntervals;
        }

        public int DatasetSize { get; protected set; }public int Attributes { get; protected set; }public int InputAttributes { get; protected set; }public int OutputAttributes { get; protected set; }public int OutputIntervals { get; protected set; }public string Filename { get; protected set; }
        public double[][] Dataset { get; set; }public int[] LingvisticAttribute { get; set; }public double InitialError { get; set; }public void ClearDataset()
        {
            DatasetSize = 0;InputAttributes = 0;OutputAttributes = 0;OutputIntervals = 0;InitialError = -1;Filename = "text.txt";Dataset = null;LingvisticAttribute = null;
        }
        public abstract int InitializeDataset();
        public void NormalizeDataset()
        {
            var min = new double[InputAttributes];
            var max = new double[InputAttributes];

            for (var i = 0; i < InputAttributes; i++)
                if (LingvisticAttribute[i] == 0)
                {
                    min[i] = Dataset[0][i];
                    max[i] = Dataset[0][i];
                    for (var k = 1; k < DatasetSize; k++)
                    {
                        if (Dataset[k][i] < min[i]) min[i] = Dataset[k][i];
                        if (Dataset[k][i] > max[i]) max[i] = Dataset[k][i];
                    }
                    for (var k = 0; k < DatasetSize; k++)
                        Dataset[k][i] = (Dataset[k][i] - min[i])/(max[i] - min[i]);
                }
        }

        
        public double ComputeInitialError()
        {
            InitialError = 0.0;int max;
                var classes = new int[OutputIntervals];

                for (var i = 0; i < OutputIntervals; i++)
                    classes[i] = 0;

                for (var x = 0; x < DatasetSize; x++)
                    classes[Convert.ToInt32(Dataset[x][InputAttributes])]++;

                max = classes[0];
                for (var i = 0; i < OutputIntervals; i++)
                    if (classes[i] > max)
                        max = classes[i];
                InitialError = 1 - max*1.0f/DatasetSize;
          
            return InitialError;
        }

        public override string ToString()
        {
            var basic = "Dataset Size: \t\t\t\t" + DatasetSize + "\n" +
                        "Attributes: \t\t\t\t\t" + Attributes + "\n" +
                        "Input Attributes: \t\t" + InputAttributes + "\n" +
                        "Output Attributes: \t\t" + OutputAttributes + "\n" +
                        "Output Intervals: \t\t" + OutputIntervals + "\n" +
                        "Initial Error: \t\t\t\t" + InitialError + "\n" +
                        "Filename: \t\t\t\t\t\t" + Filename + "\n";

            basic += "LingvisticAttribute: \t";
            for (var i = 0; i < LingvisticAttribute.Length; i++)
                basic += LingvisticAttribute[i] + "\t";
            basic += "\n";


            basic += "Dataset:\n";
            foreach (var x in Dataset)
            {
                foreach (var y in x)
                    basic += Math.Round(y, 4).ToString("0.0000") + "\t";
                basic += "\n";
            }
            return basic;
        }

        public void WriteToFile(string filename = "dataset.txt")
        {
            using (var w = new StreamWriter(filename))
            {
                foreach (var x in Dataset)
                {
                    foreach (var y in x)
                        w.Write(Math.Round(y, 4).ToString("0.0000") + "\t");
                    w.WriteLine();
                }
            }
        }

        public void WriteInfoToFile(string filename = "dataset-info.txt")
        {
            using (var w = new StreamWriter(filename))
            {
                w.WriteLine(ToString());
            }
        }

        public void RandomizeArray()
        {
            new Random().Shuffle(Dataset);
        }


        public void ShrinkDataset(int size)
        {
            if (size < DatasetSize)
            {
                var shrinked = new double[size][];
                DatasetSize = size;

                for (var i = 0; i < Dataset.Length; i++)
                    if (i < size)
                        shrinked[i] = Dataset[i];
                Dataset = shrinked;
            }
            ComputeInitialError();
            NormalizeDataset();
        }
    }

    internal static class RandomExtensions
    {
        public static void Shuffle<T>(this Random rng, T[] array){var n = array.Length;while (n > 1){var k = rng.Next(n--);var temp = array[n];array[n] = array[k];array[k] = temp;}
        }
    }
}

File: zdrojove kody v c++/DataSets.cpp
#include "DataSets.h"
#include <ctime>


namespace Datasets
{
	DataSets::DataSets(int datasetSize, int inputAttributes, int outputIntervals, const std::wstring& filename)
	{
		srand(time(NULL));
		setDatasetSize(datasetSize);
		setAttributes(inputAttributes + 1);
		setInputAttributes(inputAttributes);
		setOutputAttributes(1);
		setOutputIntervals(outputIntervals);
		setFilename(filename);
		setInitialError(-1);
		setDataset(std::vector<std::vector<double>>(datasetSize));
		for (int i = 0; i < datasetSize; i++)
		{
			getDataset()[i] = std::vector<double>(getAttributes());
		}

		setLingvisticAttribute(std::vector<int>(getAttributes()));
		for (int i = 0; i < getLingvisticAttribute().size(); i++)
		{
			getLingvisticAttribute()[i] = 0;
		}
		getLingvisticAttribute()[inputAttributes] = outputIntervals;
	}

	int DataSets::getDatasetSize() const
	{
		return _datasetSize;
	}

	void DataSets::setDatasetSize(const int& value)
	{
		_datasetSize = value;
	}

	int DataSets::getAttributes() const
	{
		return _attributes;
	}

	void DataSets::setAttributes(const int& value)
	{
		_attributes = value;
	}

	int DataSets::getInputAttributes() const
	{
		return _inputAttributes;
	}

	void DataSets::setInputAttributes(const int& value)
	{
		_inputAttributes = value;
	}

	int DataSets::getOutputAttributes() const
	{
		return _outputAttributes;
	}

	void DataSets::setOutputAttributes(const int& value)
	{
		_outputAttributes = value;
	}

	int DataSets::getOutputIntervals() const
	{
		return _outputIntervals;
	}

	void DataSets::setOutputIntervals(const int& value)
	{
		_outputIntervals = value;
	}

	std::wstring DataSets::getFilename() const
	{
		return _filename;
	}

	void DataSets::setFilename(const std::wstring& value)
	{
		_filename = value;
	}

	std::vector<std::vector<double>> DataSets::getDataset() const
	{
		return _dataset;
	}

	void DataSets::setDataset(const std::vector<std::vector<double>>& value)
	{
		_dataset = value;
	}

	std::vector<int> DataSets::getLingvisticAttribute() const
	{
		return _lingvisticAttribute;
	}

	void DataSets::setLingvisticAttribute(const std::vector<int>& value)
	{
		_lingvisticAttribute = value;
	}

	double DataSets::getInitialError() const
	{
		return _initialError;
	}

	void DataSets::setInitialError(const double& value)
	{
		_initialError = value;
	}

	void DataSets::ClearDataset()
	{
		setDatasetSize(0);
		setInputAttributes(0);
		setOutputAttributes(0);
		setOutputIntervals(0);
		setInitialError(-1);
		setFilename(L"text.txt");
		getDataset().clear();
		getLingvisticAttribute().clear();
	}

	void DataSets::NormalizeDataset()
	{
		auto min = std::vector<double>(getInputAttributes());
		auto max = std::vector<double>(getInputAttributes());

		for (int i = 0; i < getInputAttributes(); i++)
		{
			if (getLingvisticAttribute()[i] == 0)
			{
				min[i] = getDataset()[0][i];
				max[i] = getDataset()[0][i];
				for (int k = 1; k < getDatasetSize(); k++)
				{
					if (getDataset()[k][i] < min[i])
					{
						min[i] = getDataset()[k][i];
					}
					if (getDataset()[k][i] > max[i])
					{
						max[i] = getDataset()[k][i];
					}
				}
				for (int k = 0; k < getDatasetSize(); k++)
				{
					getDataset()[k][i] = (getDataset()[k][i] - min[i]) / (max[i] - min[i]);
				}
			}
		}
	}

	double DataSets::ComputeInitialError()
	{
		setInitialError(0.0);
		int max;
		auto classes = std::vector<int>(getOutputIntervals());

		for (int i = 0; i < getOutputIntervals(); i++)
		{
			classes[i] = 0;
		}

		for (int x = 0; x < getDatasetSize(); x++)
		{
			classes[static_cast<int>(getDataset()[x][getInputAttributes()])]++;
		}

		max = classes[0];
		for (int i = 0; i < getOutputIntervals(); i++)
		{
			if (classes[i] > max)
			{
				max = classes[i];
			}
		}
		setInitialError(1 - max * 1.0f / getDatasetSize());

		return getInitialError();
	}


	void DataSets::ShrinkDataset(int size)
	{
		if (size < getDatasetSize())
		{
			auto shrinked = std::vector<std::vector<double>>(size);
			setDatasetSize(size);

			for (int i = 0; i < getDataset().size(); i++)
			{
				if (i < size)
				{
					shrinked[i] = getDataset()[i];
				}
			}
			setDataset(shrinked);
		}
		ComputeInitialError();
		NormalizeDataset();
	}

	template <typename T>
	void RandomExtensions::Shuffle(std::vector<T>& array_Renamed)
	{
		auto n = array_Renamed.size();
		while (n > 1)
		{
			auto k = rand % (n--);
			auto temp = array_Renamed[n];
			array_Renamed[n] = array_Renamed[k];
			array_Renamed[k] = temp;
		}
	}
}
